src/PFire.Core/Services/PFireDatabase.cs

```diff
@@ -0,0 +1,258 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.DependencyInjection;
+using PFire.Core.Models;
+using PFire.Data.Entities;
+using PFire.Data.Services;
+
+namespace PFire.Core.Services
+{
+    internal interface IPFireDatabase
+    {
+        Task<UserModel> InsertUser(string username, string password, string salt);
+        Task InsertMutualFriend(UserModel me, UserModel them);
+        Task InsertFriendRequest(UserModel me, UserModel them, string message);
+        Task<UserModel> QueryUser(string username);
+        Task<List<UserModel>> QueryUsers(string username);
+        Task<List<UserModel>> QueryFriends(UserModel user);
+        Task<List<FriendRequestModel>> QueryPendingFriendRequestsSelf(UserModel user);
+        Task<List<FriendRequestModel>> QueryPendingFriendRequests(UserModel otherUser);
+        Task DeletePendingFriendRequest(UserModel me, params FriendRequestModel[] thems);
+        Task UpdateNickname(UserModel user, string nickname);
+    }
+
+    internal class PFireDatabase : IPFireDatabase
+    {
+        private readonly IServiceProvider _serviceProvider;
+
+        public PFireDatabase(IServiceProvider serviceProvider)
+        {
+            _serviceProvider = serviceProvider;
+        }
+
+        public async Task<UserModel> InsertUser(string username, string password, string salt)
+        {
+            using var scope = _serviceProvider.CreateScope();
+            var databaseContext = scope.ServiceProvider.GetRequiredService<IDatabaseContext>();
```

Comment:
We _can_, but I chose to do it this way because this service is injected into the PFireServer "globally". With this style, we're not keeping around a single DbContext for the life of the application, but connecting only when we want to interact with the database.

If we get around to creating a scope-per-message, then we can inject the DbContext normally.

Let me know if that's okay for now.

Reply:
Yeah that makes sense!